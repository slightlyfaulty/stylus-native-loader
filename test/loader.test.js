import path from 'path'

import stylus from 'stylus'
import nib from 'nib'

import {compile} from './helpers/compiler'

function check(stats, outputIndex = 0) {
	expect(stats.compilation.warnings).toEqual([])
	expect(stats.result.output[outputIndex]).toMatchSnapshot('output')

	if (process.env.DEBUG_OUTPUT) {
		console.debug(stats.result.output[outputIndex])
	}
}

describe("loader", () => {
	it("should work", async () => {
		check(await compile('basic.styl'))
	})

	it("should generate source maps accordingly", async () => {
		// generate source map with full source content (devtool)
		check(await compile('basic.styl', {}, {
			devtool: 'source-map'
		}), 1)

		// generate source map with full source content (sourceMap)
		check(await compile('basic.styl', {
			sourceMap: true
		}, {
			devtool: 'eval-cheap-source-map'
		}), 1)

		// generate source map with without source content
		check(await compile('basic.styl', {
			sourcemap: {
				content: false
			},
		}, {
			devtool: 'source-map'
		}), 1)

		// don't generate source map (devtool)
		check(await compile('basic.styl', {}, {
			devtool: 'eval-cheap-source-map',
		}), 1)

		// don't generate source map (sourceMap)
		check(await compile('basic.styl', {
			sourceMap: false,
		}, {
			devtool: 'source-map'
		}), 1)
	})

	it("should allow using plugins ('use' option)", async () => {
		// import in stylesheet
		check(await compile('use.styl', {
			use: nib()
		}))

		// import in loader
		check(await compile('nibTest.styl', {
			use: nib(),
			import: 'nib',
		}))

		// use string
		check(await compile('use.styl', {
			use: 'nib',
		}))

		// use string and import
		check(await compile('nibTest.styl', {
			use: 'nib',
			import: 'nib',
		}))
	})

	it("should allow adding include paths ('include' option)", async () => {
		check(await compile('include.styl', {
			include: path.join(__dirname, 'fixtures/imports')
		}))
	})

	it("should allow importing files ('import' option)", async () => {
		check(await compile('import.styl', {
			import: path.join(__dirname, 'fixtures/imports/a.styl')
		}))
	})

	it("should allow defining variables/functions ('define' option)", async () => {
		const define = {
			'$padding': new stylus.nodes.Unit(10, 'px'),
			'$object': {colors: {primary: '#123456'}},
			'double': n => n.operate('+', n),
		}

		// objects converted to hashes by default
		check(await compile('define.styl', {
			define
		}))

		// objects converted to lists (aka expressions)
		check(await compile('define.styl', {
			define,
			defineRaw: false,
		}))
	})

	it("should resolve urls accordingly ('resolveUrl' option)", async () => {
		// resolve urls
		check(await compile('resolveUrl.styl', {
			resolveUrl: true
		}))

		// resolve urls without checking file existence
		check(await compile('resolveUrl.styl', {
			resolveUrl: 'nocheck'
		}))

		// don't resolve urls
		check(await compile('resolveUrl.styl', {
			resolveUrl: false
		}))
	})

	it("should resolve aliases", async () => {
		const alias = {
			'imp-a$': path.join(__dirname, 'fixtures/imports/a'),
			'~b': path.join(__dirname, 'fixtures/imports/b.styl'),
			'imp': path.join(__dirname, 'fixtures/imports'),
		}

		// resolve aliases set in webpack config
		check(await compile('alias.styl', {}, {
			resolve: {
				alias
			},
		}))

		// resolve aliases set in loader options
		check(await compile('alias.styl', {
			alias
		}))
	})

	it("should resolve tilde paths", async () => {
		check(await compile('tilde.styl'))
	})

	it("should allow toggling CSS file imports ('includeCSS' option)", async () => {
		const alias = {
			'b': path.join(__dirname, 'fixtures/css/b.css'),
		}

		// don't include imported CSS in output by default
		check(await compile('importCSS.styl', {
			alias,
		}))

		// include imported CSS in output
		check(await compile('importCSS.styl', {
			alias,
			includeCSS: true,
		}))
	})

	it("should allow toggling vendor prefixes ('vendors' option)", async () => {
		// vendors disabled by default
		check(await compile('vendors.styl'))

		// vendors enabled
		check(await compile('vendors.styl', {
			vendors: true,
		}))
	})

	it("should trigger 'beforeCompile' callback", async () => {
		check(await compile('include.styl', {
			beforeCompile(renderer, context, options) {
				renderer.include(path.join(__dirname, 'fixtures/imports'))

				if (!context.rootContext) {
					throw new Error(`beforeCompile() - invalid "context" parameter.`)
				}

				if (!options.beforeCompile) {
					throw new Error(`beforeCompile() - invalid "options" parameter.`)
				}
			}
		}))
	})

	it("should resolve relative imports", async () => {
		check(await compile('relative/index.styl'))
	})

	it("should resolve glob imports", async () => {
		check(await compile('globs/index.styl', {
			resolveUrl: true, // test with Stylus resolveURL glob bug
		}))
	})

	it("should work with vue component files", async () => {
		const VueLoaderPlugin = require('vue-loader/lib/plugin')

		check(await compile('vue/index.js', {}, {
			module: {
				rules: [
					{
						test: /\.vue$/,
						loader: 'vue-loader',
					},
				],
			},
			plugins: [
				new VueLoaderPlugin()
			],
		}))
	})
})
