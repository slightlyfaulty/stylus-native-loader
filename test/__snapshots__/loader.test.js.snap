// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`loader should allow adding include paths ('include' option): output 1`] = `
".a {
  color: #aaa;
}
.b {
  color: #bbb;
}
.c {
  color: #ccc;
}
"
`;

exports[`loader should allow defining variables/functions ('define' option): output 1`] = `
".define {
  padding: 10px;
  margin: 20px;
}
.hash {
  color: '#123456';
}
"
`;

exports[`loader should allow defining variables/functions ('define' option): output 2`] = `
".define {
  padding: 10px;
  margin: 20px;
}
.expression {
  color: '#123456';
}
"
`;

exports[`loader should allow importing files ('import' option): output 1`] = `
".a {
  color: #aaa;
}
.b {
  color: #bbb;
}
.c {
  color: #ccc;
}
"
`;

exports[`loader should allow toggling CSS file imports ('includeCSS' option): output 1`] = `
"@import url("https://fonts.googleapis.com/css2?family=Open+Sans:wght@300;400;600&display=swap");
@import 'css/a.css';
@import 'b';
"
`;

exports[`loader should allow toggling CSS file imports ('includeCSS' option): output 2`] = `
"@import url("https://fonts.googleapis.com/css2?family=Open+Sans:wght@300;400;600&display=swap");
.a {
	color: #aaa;
}
.b {
	color: #bbb;
}
"
`;

exports[`loader should allow toggling vendor prefixes ('vendors' option): output 1`] = `
".nib {
  background: linear-gradient(to bottom, #fff, #000);
  position: fixed;
  bottom: 0;
  right: 0;
}
@keyframes slidein {
  from {
    transform: translateX(0%);
  }
  to {
    transform: translateX(100%);
  }
}
"
`;

exports[`loader should allow toggling vendor prefixes ('vendors' option): output 2`] = `
".nib {
  background: -webkit-linear-gradient(top, #fff, #000);
  background: -moz-linear-gradient(top, #fff, #000);
  background: -o-linear-gradient(top, #fff, #000);
  background: -ms-linear-gradient(top, #fff, #000);
  background: linear-gradient(to bottom, #fff, #000);
  position: fixed;
  bottom: 0;
  right: 0;
}
@-moz-keyframes slidein {
  from {
    -webkit-transform: translateX(0%);
    -moz-transform: translateX(0%);
    -o-transform: translateX(0%);
    -ms-transform: translateX(0%);
    transform: translateX(0%);
  }
  to {
    -webkit-transform: translateX(100%);
    -moz-transform: translateX(100%);
    -o-transform: translateX(100%);
    -ms-transform: translateX(100%);
    transform: translateX(100%);
  }
}
@-webkit-keyframes slidein {
  from {
    -webkit-transform: translateX(0%);
    -moz-transform: translateX(0%);
    -o-transform: translateX(0%);
    -ms-transform: translateX(0%);
    transform: translateX(0%);
  }
  to {
    -webkit-transform: translateX(100%);
    -moz-transform: translateX(100%);
    -o-transform: translateX(100%);
    -ms-transform: translateX(100%);
    transform: translateX(100%);
  }
}
@-o-keyframes slidein {
  from {
    -webkit-transform: translateX(0%);
    -moz-transform: translateX(0%);
    -o-transform: translateX(0%);
    -ms-transform: translateX(0%);
    transform: translateX(0%);
  }
  to {
    -webkit-transform: translateX(100%);
    -moz-transform: translateX(100%);
    -o-transform: translateX(100%);
    -ms-transform: translateX(100%);
    transform: translateX(100%);
  }
}
@keyframes slidein {
  from {
    -webkit-transform: translateX(0%);
    -moz-transform: translateX(0%);
    -o-transform: translateX(0%);
    -ms-transform: translateX(0%);
    transform: translateX(0%);
  }
  to {
    -webkit-transform: translateX(100%);
    -moz-transform: translateX(100%);
    -o-transform: translateX(100%);
    -ms-transform: translateX(100%);
    transform: translateX(100%);
  }
}
"
`;

exports[`loader should allow using plugins ('use' option): output 1`] = `
".nib {
  background: linear-gradient(to bottom, #fff, #000);
  position: fixed;
  bottom: 0;
  right: 0;
}
"
`;

exports[`loader should allow using plugins ('use' option): output 2`] = `
".nib {
  background: linear-gradient(to bottom, #fff, #000);
  position: fixed;
  bottom: 0;
  right: 0;
}
"
`;

exports[`loader should allow using plugins ('use' option): output 3`] = `
".nib {
  background: linear-gradient(to bottom, #fff, #000);
  position: fixed;
  bottom: 0;
  right: 0;
}
"
`;

exports[`loader should allow using plugins ('use' option): output 4`] = `
".nib {
  background: linear-gradient(to bottom, #fff, #000);
  position: fixed;
  bottom: 0;
  right: 0;
}
"
`;

exports[`loader should generate source maps accordingly: output 1`] = `
{
  "mappings": "AAKA;EAFC,OAAM,oBAAN;;ACHD;EDGC,OAAM,KAAN;;AEHD;EFGC,OAAM,KAAN;;AGHD;EHGC,OAAM,KAAN",
  "names": [],
  "sourceRoot": "/x/packages/stylus-native-loader/test/fixtures",
  "sources": [
    "test/fixtures/basic.styl",
    "test/fixtures/imports/a.styl",
    "test/fixtures/imports/b.styl",
    "test/fixtures/imports/c.styl",
  ],
  "sourcesContent": [
    "$color = red

color(color)
	color complement(color)

body
	color rgba($color, .5)

@require 'imports'",
    ".a
	color #aaa",
    ".b
	color #bbb",
    ".c
	color #ccc",
  ],
  "version": 3,
}
`;

exports[`loader should generate source maps accordingly: output 2`] = `
{
  "mappings": "AAKA;EAFC,OAAM,oBAAN;;ACHD;EDGC,OAAM,KAAN;;AEHD;EFGC,OAAM,KAAN;;AGHD;EHGC,OAAM,KAAN",
  "names": [],
  "sourceRoot": "/x/packages/stylus-native-loader/test/fixtures",
  "sources": [
    "test/fixtures/basic.styl",
    "test/fixtures/imports/a.styl",
    "test/fixtures/imports/b.styl",
    "test/fixtures/imports/c.styl",
  ],
  "sourcesContent": [
    "$color = red

color(color)
	color complement(color)

body
	color rgba($color, .5)

@require 'imports'",
    ".a
	color #aaa",
    ".b
	color #bbb",
    ".c
	color #ccc",
  ],
  "version": 3,
}
`;

exports[`loader should generate source maps accordingly: output 3`] = `
{
  "mappings": "AAKA;EAFC,OAAM,oBAAN;;ACHD;EDGC,OAAM,KAAN;;AEHD;EFGC,OAAM,KAAN;;AGHD;EHGC,OAAM,KAAN",
  "names": [],
  "sourceRoot": "/x/packages/stylus-native-loader/test/fixtures",
  "sources": [
    "test/fixtures/basic.styl",
    "test/fixtures/imports/a.styl",
    "test/fixtures/imports/b.styl",
    "test/fixtures/imports/c.styl",
  ],
  "version": 3,
}
`;

exports[`loader should generate source maps accordingly: output 4`] = `undefined`;

exports[`loader should generate source maps accordingly: output 5`] = `undefined`;

exports[`loader should resolve aliases: output 1`] = `
".a {
  color: #aaa;
}
.b {
  color: #bbb;
}
.c {
  color: #ccc;
}
"
`;

exports[`loader should resolve aliases: output 2`] = `
".a {
  color: #aaa;
}
.b {
  color: #bbb;
}
.c {
  color: #ccc;
}
"
`;

exports[`loader should resolve glob imports: output 1`] = `
".a {
  color: #aaa;
}
.b {
  color: #bbb;
}
.c {
  color: #ccc;
}
"
`;

exports[`loader should resolve relative imports: output 1`] = `
".a-color {
  color: #aaa;
}
.b-color {
  color: #bbb;
}
.c-color {
  color: #ccc;
}
"
`;

exports[`loader should resolve tilde paths: output 1`] = `
".nib {
  background: linear-gradient(to bottom, #fff, #000);
  position: fixed;
  bottom: 0;
  right: 0;
}
"
`;

exports[`loader should resolve urls accordingly ('resolveUrl' option): output 1`] = `
".test {
  background-image: url("urls/img.png");
}
.fake {
  background-image: url("fake.png");
}
.nested1 {
  background-image: url("urls/nested/img.png");
}
.nested2 {
  background-image: url("urls/nested/img.png");
}
.nested3 {
  background-image: url("urls/img.png");
}
"
`;

exports[`loader should resolve urls accordingly ('resolveUrl' option): output 2`] = `
".test {
  background-image: url("urls/img.png");
}
.fake {
  background-image: url("urls/fake.png");
}
.nested1 {
  background-image: url("urls/nested/img.png");
}
.nested2 {
  background-image: url("urls/nested/img.png");
}
.nested3 {
  background-image: url("urls/img.png");
}
"
`;

exports[`loader should resolve urls accordingly ('resolveUrl' option): output 3`] = `
".test {
  background-image: url("img.png");
}
.fake {
  background-image: url("fake.png");
}
.nested1 {
  background-image: url("nested/img.png");
}
.nested2 {
  background-image: url("img.png");
}
.nested3 {
  background-image: url("../img.png");
}
"
`;

exports[`loader should trigger 'beforeCompile' callback: output 1`] = `
".a {
  color: #aaa;
}
.b {
  color: #bbb;
}
.c {
  color: #ccc;
}
"
`;

exports[`loader should work with vue component files: output 1`] = `
"body {
  color: rgba(0,255,255,0.5);
}
.a {
  color: #aaa;
}
.b {
  color: #bbb;
}
.c {
  color: #ccc;
}
h1 {
  color: #c04f82;
}
"
`;

exports[`loader should work: output 1`] = `
"body {
  color: rgba(0,255,255,0.5);
}
.a {
  color: #aaa;
}
.b {
  color: #bbb;
}
.c {
  color: #ccc;
}
"
`;
